File: AgqrPlayer4Tv/AgqrPlayer4Tv/Components/Player/RtmpDataSource.cs
using System;
using Com.Google.Android.Exoplayer;
using Com.Google.Android.Exoplayer.Upstream;
using Net.Butterflytv.Rtmp_client;

namespace AgqrPlayer4Tv.Components.Player
{
    public class RtmpDataSource : Java.Lang.Object, IUriDataSource
    {
        private RtmpClient _rtmpClient;
        private bool _isOpened = false;

        public RtmpDataSource()
        { }

        public RtmpDataSource(string uri)
        {
            this.Uri = uri;
        }

        public void Close()
        {
            this._rtmpClient.Close();
            this._rtmpClient = null;
            this._isOpened = false;
        }

        public long Open(DataSpec p0)
        {
            if (this._isOpened)
            {
                this.Close();
            }

            this.Uri = p0.Uri.ToString();
            this._rtmpClient = new RtmpClient();
            this._rtmpClient.Open(p0.Uri.ToString(), false);
            this._isOpened = true;
            return C.LengthUnbounded;
        }

        public int Read(byte[] p0, int p1, int p2)
        {
            if (this._rtmpClient.IsConnected() == 0) return 0;
            return this._rtmpClient.Read(p0, p1, p2);
        }

        public string Uri { get; private set; }
    }
}

File: AgqrPlayer4Tv/app/src/main/java/org/misuzilla/agqrplayer4tv/infrastracture/exoplayer/RtmpDataSource.kt
package org.misuzilla.agqrplayer4tv.infrastracture.exoplayer

import android.net.Uri
import com.google.android.exoplayer2.C
import com.google.android.exoplayer2.upstream.DataSource
import com.google.android.exoplayer2.upstream.DataSpec
import com.google.android.exoplayer2.upstream.TransferListener
import net.butterflytv.rtmp_client.RtmpClient

/**
 * Created by Tomoyo on 12/11/2016.
 */
class RtmpDataSource : DataSource {
    private val rtmpClient = RtmpClient()
    private var uri: Uri? = null

    override fun open(dataSpec: DataSpec): Long {
        uri = dataSpec.uri
        rtmpClient.open(uri?.toString(), false)

        return C.LENGTH_UNSET.toLong()
    }

    override fun getUri(): Uri {
        return uri!!
    }

    override fun close() {
        rtmpClient.close()
    }

    override fun addTransferListener(transferListener: TransferListener?) {
    }

    override fun read(buffer: ByteArray?, offset: Int, readLength: Int): Int {
        if (!rtmpClient.isConnected) return 0
        return rtmpClient.read(buffer, offset, readLength)
    }

}